src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
+        // https://stackoverflow.com/a/37528114 provides an explanation for why.
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (_decompressionStream is null)
+            {
+                int firstByte = _stream.PeekFirstByte();
+                _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+            }
+
+            return _decompressionStream.Read(buffer);
+        }
+
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            if (_decompressionStream is null)
+            {
+                return CreateAndReadAsync(this, buffer, cancellationToken);
+
+                static async ValueTask<int> CreateAndReadAsync(ZLibOrDeflateStream thisRef, Memory<byte> buffer, CancellationToken cancellationToken)
+                {
+                    int firstByte = await thisRef._stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                    thisRef._decompressionStream = CreateDecompressionStream(firstByte, thisRef._stream);
+                    return await thisRef._decompressionStream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return _decompressionStream.ReadAsync(buffer, cancellationToken);
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            ValidateCopyToArguments(destination, bufferSize);
+            return Core(destination, bufferSize, cancellationToken);
+            async Task Core(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                if (_decompressionStream is null)
+                {
+                    int firstByte = await _stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                    _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+                }
+
+                await _decompressionStream.CopyToAsync(destination, bufferSize, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count) => _stream.Read(buffer, offset, count);
+        public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+        public override void SetLength(long value) => _stream.SetLength(value);
+        public override void Flush() => _stream.Flush();
+
+        private static Stream CreateDecompressionStream(int firstByte, Stream stream) =>
+            (firstByte & 0xF) == 8 ?
+                new ZLibStream(stream, CompressionMode.Decompress) :
+                new DeflateStream(stream, CompressionMode.Decompress);
+
+        // As the name suggests, this is just a Stream that allows peeking at the first byte.
+        private sealed class PeekFirstByteReadStream : Stream
```

Comment:
Instead of inventing PeekFirstByteReadStream would it work to just use a StreamReader passing in a buffer size of 1?

Reply:
@stephentoub curious whether you see another possible trick to simplify here.

Reply:
actually, why not just call ReadByte() on the Stream you're given?

Reply:
Synchronous reads aren't allowed by default in aspnetcore.

Reply:
I think I see what you're saying, `PeekFirstByteReadStream` shouldn't be needed I think.

Reply:
Why shouldn't it be needed? Why need to read the first byte in order to decide which wrapper to use, but that wrapper also needs to read the first byte, so the peek stream allows accessing the byte without consuming it. It wouldn't be needed if seeking were possible but it's not on a network stream. What is the alternative being suggested? If it's too employ a buffer, that's effectively what this is doing. An alternative would be to wrap the input in a pipe reader, peek on that without consuming, and then wrap that reader in a stream that's passed to the decompression stream.

Reply:
I overlooked that it's a peek, meaning the concrete streams get to see the full content.

There's still some tidying that can be done though.

Reply:
> There's still some tidying that can be done though.

Such as?

Reply:
Several of my other comments?

Reply:
Is StreamReader an alternative to PeekFirstByteReadStream ? 
```c#
using (var sr = new StreamReader(stream, encoding: null, detectEncodingFromByteOrderMarks: false, bufferSize: 1, leaveOpen: true))
  byte = sr.Peek();
```
it allocates a little more (eg., a Decoder) but eliminates the PeekFirstByteReadStream?

Reply:
> Is StreamReader an alternative to PeekFirstByteReadStream ?

No. You need to be able to construct a DeflateStream or ZLibStream that's handed all the data.  The only way that StreamReader example would work is if the `stream` is seekable and you could seek back to position 0 after reading, but if you can seek like that, you don't need the StreamReader at all.

Reply:
Ah -- essentially, because StreamReader's Peek() does not call the wrapped stream Peek(). It calls Read() (albeit only 1 byte in this case but it advances)